lib/services/local/users.go

```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
 		return trace.Wrap(err)
 	}
 
-	err = s.UpsertPasswordHash(user, hash)
-	if err != nil {
+	if err := s.upsertPasswordHash(user, hash); err != nil {
 		return trace.Wrap(err)
 	}
 
+	_, err = s.UpdateAndSwapUser(
+		context.TODO(),
+		user,
+		false, /*withSecrets*/
+		func(u types.User) (bool, error) {
+			u.SetPasswordState(types.PasswordState_PASSWORD_STATE_SET)
+			return true, nil
+		})
+	if err != nil {
+		// Don't let the password state flag change fail the entire operation.
+		s.log.WithError(err).Info("Failed to set password state")
```

Comment:
Bump to warning, log user name?

```suggestion
		s.log.
			WithError(err).
			WithField("User", user).
			Warn("Failed to set user password state")
```